pkg/secrethub/internals/http/client.go

```diff
@@ -245,11 +245,40 @@ func (c *Client) DeleteRepo(namespace, repoName string) error {
 }
 
 // AuditRepo gets the audit events for a given repo.
-func (c *Client) AuditRepo(namespace, repoName string, subjectTypes api.AuditSubjectTypeList) ([]*api.Audit, error) {
-	out := []*api.Audit{}
-	rawURL := fmt.Sprintf(pathRepoEvents+"?subject_types=%s", c.base, namespace, repoName, subjectTypes.Join(","))
-	err := c.get(rawURL, true, &out)
-	return out, errio.Error(err)
+func (c *Client) AuditRepo(namespace, repoName string, subjectTypes api.AuditSubjectTypeList) (*Paginator, error) {
+	requestURL, err := url.Parse(fmt.Sprintf(pathRepoEvents, c.base, namespace, repoName))
+	if err != nil {
+		return nil, fmt.Errorf("invalid URL: %s", err)
+	}
+
+	if len(subjectTypes) > 0 {
+		q := requestURL.Query()
+		q.Set("subject_types", subjectTypes.Join(","))
+		requestURL.RawQuery = q.Encode()
+	}
+
+	return c.auditPage(*requestURL), nil
+}
+
+func (c *Client) auditPage(url url.URL) *Paginator {
+	events := make([]api.Audit, 50)
+	return NewPaginator(
+		url,
+		&events,
+		func(vs interface{}) string {
+			events := *vs.(*[]api.Audit)
+			return events[len(events)-1].EventID.String()
```

Comment:
What happens when `len(events)` is `0`? And is that a case we need to account for?

Reply:
This function should never be called on an empty list. This is ensured by this check:
https://github.com/secrethub/secrethub-go/pull/144/files/83e130dc31279abcb00411bf94e53acb412c5f2f#diff-cc52c786e26f396dead3234863b06611R34

Alternatively, we could make the function return an error and only add the request parameter when the function returns no error (`nil`). That would mean we have to add this zero length check for every type of paginator though (auditevents, users etc.), instead of defining it once in the paginator.

When we add the link header to the response, we can drop this function altogether, as we can then get the next page from the response.

Reply:
These checks are now moved together.